File: fortran/numbers/src/numbers_iterative.f90
module numbers_iterative
contains

  function fibonacci(N) result(res)
    integer, intent(in) :: N
    integer :: res
    res = fibonacci_iter(N, 0, 1)
  end function fibonacci

  recursive function fibonacci_iter(N, Acc2, Acc1) result(res)
    integer, intent(in) :: N, Acc2, Acc1
    integer :: res
    if (N <= 1) then
      res = N
    else if (N == 2) then
      res = Acc1 + Acc2
    else
      res = fibonacci_iter(N - 1, Acc1, Acc1 + Acc2)
    end if
  end function fibonacci_iter

  function factorial(N) result(res)
    integer, intent(in) :: N
    res = factorial_iter(N, 1)
  end function factorial

  recursive function factorial_iter(N, Accumulator) result(res)
    integer, intent(in) :: N, Accumulator
    integer :: res
    if (N == 1) then
      res = Accumulator
    else
      res = factorial_iter(N - 1, N * Accumulator)
    end if
  end function factorial_iter

  function sum_numbers(N) result(res)
    integer, intent(in) :: N
    res = sum_numbers_iter(N, 0)
  end function sum_numbers

  recursive function sum_numbers_iter(N, Accumulator) result(res)
    integer, intent(in) :: N, Accumulator
    integer :: res
    if (N == 0) then
      res = Accumulator
    else
      res = sum_numbers_iter(N - 1, N + Accumulator)
    end if
  end function sum_numbers_iter

end module numbers_iterative

File: julia/numbers/src/NumbersIterative.jl
#!/usr/bin/julia

module NumbersIterative
export fibonacci, factorial_rec, sum_numbers

fibonacci(n::Int) :: Int = fibonacci_iter(n, 0, 1)
fibonacci_iter(n::Int, acc2::Int, acc1::Int) :: Int = 
    if n <= 0 acc2
    elseif n <= 2 acc2 + acc1
    else fibonacci_iter(n - 1, acc1, acc2 + acc1)
    end
factorial_rec(n::Int) :: Int = factorial_iter(n, 1)
factorial_iter(n::Int, acc::Int) :: Int = n == 0 ?  acc : factorial_iter(n - 1, n * acc)
sum_numbers(n::Int) :: Int = sum_numbers_iter(n, 0)
sum_numbers_iter(n::Int, acc::Int) :: Int = n == 0 ? acc : sum_numbers_iter(n - 1, n + acc)
end